File: 2022-7-17/565.数组嵌套.go
package main

import "fmt"


func arrayNesting(nums []int) int {
	ret := 0
	exist := make([]bool, len(nums))

	for i := range nums {
		cnt := 0
		for {
			if exist[i] == true {
				break
			}
			exist[i] = true
			i = nums[i]
			cnt++
		}

		if cnt > ret {
			ret = cnt
		}
	}

	return ret
}

func main() {
	A := []int{5, 4, 0, 3, 1, 6, 2}

	ret := arrayNesting(A)

	fmt.Printf("ret is %d\n", ret)
}

File: 2022-7-17/565.数组嵌套.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define MAX(a, b) ((a) > (b) ? (a) : (b))

int array_nesting(int* nums, int numsSize){
    int ans = 0;
    bool *vis = (bool *)malloc(sizeof(bool) * numsSize);
    for (int i = 0; i < numsSize; ++i) {
        int cnt = 0;
        memset(vis, 0, sizeof(bool) * numsSize);
        while (!vis[i]) {
            vis[i] = true;
            i = nums[i];
            ++cnt;
        }
        ans = MAX(ans, cnt);
    }
    free(vis);
    return ans;
}

int main()
{
    int ret;
    int V[7] = {5, 4, 0, 3, 1, 6, 2};

    ret = array_nesting(V, 7);
    printf("ret is %d\n", ret);

    return 0;
}